### PR TITLE
Build 64bit binaries for Android

### LINKED ISSUE #2814: Support third-party 64-bit libraries on Android
Excerpt from the Android Developers Blog article, [Get your apps ready for the 64-bit requirement](https://android-developers.googleblog.com/2019/01/get-your-apps-ready-for-64-bit.html):
*[added by @hramos]*

### The 64-bit requirement: what it means for developers

Starting **August 1, 2019**:

* All new apps and app updates that include native code are required to provide 64-bit versions in addition to 32-bit versions when publishing to Google Play.
* Extension: Google Play will continue to accept 32-bit only updates to existing games that use Unity 5.6 or older until August 2021.

Starting **August 1, 2021**:

* Google Play will stop serving apps without 64-bit versions on 64-bit capable devices, meaning they will no longer be available in the Play Store on those devices.
* This will include games built with Unity 5.6 or older.

The requirement does not apply to:

* APKs or app bundles explicitly targeting Wear OS or Android TV, which are form factors that do not currently support 64-bit code.
* APKs or app bundles that are not distributed to devices running Android 9 Pie or later.

## Original Issue

Hey, it appears that React Native on Android doesn't provide a 64-bit version of the `libreactnativejni.so` native library, which can cause compatibility issues on 64-bit devices. I ran into this while attempting to integrate React Native with a large existing application I'm developing.

Steps to reproduce:
1. Create a new application with `react-native init`.
2. In the "android/app/build.gradle" file, add `compile 'io.realm:realm-android:0.82.2'` to the end of the `dependencies {}` block. (Note that I don't think this is an issue specifically caused by Realm, it's just an example dependency with a 64-bit native library).
3. Attempt to run on a Nexus 9. 

Stack trace:

```
E/AndroidRuntime(32171): FATAL EXCEPTION: main
E/AndroidRuntime(32171): Process: com.androidemberall, PID: 32171
E/AndroidRuntime(32171): java.lang.UnsatisfiedLinkError: could find DSO to load: libreactnativejni.so
E/AndroidRuntime(32171):    at com.facebook.soloader.SoLoader.loadLibraryBySoName(SoLoader.java:213)
E/AndroidRuntime(32171):    at com.facebook.soloader.SoLoader.loadLibrary(SoLoader.java:178)
E/AndroidRuntime(32171):    at com.facebook.react.bridge.JSCJavaScriptExecutor.<clinit>(JSCJavaScriptExecutor.java:19)
...
```

I believe this to be caused by the shared library loading logic explained in [this answer](http://stackoverflow.com/a/27202958) on StackOverflow. Since Realm includes a 64-bit native binary, the system no longer automatically falls back to the 32-bit version of `libreactnativejni.so` provided by React.

This bug is blocking me from integrating React Native into my existing application, which I'm really excited to try!

### PR DESCRIPTION
**Pull request by marafat**
## Motivation
Add 64bit support to Android. Fixes #2814

**How is this different from the previous effort done in #17640 and #18754**
1. This PR uses the JSC from the SoftwareMansion's [fork](https://github.com/SoftwareMansion/jsc-android-buildscripts) [mentioned](https://github.com/facebook/react-native/issues/2814#issuecomment-391825614) under #2814 
2. It builds with NDK 16/clang
3. it updates the `APP_STL` to `c++_shared`

## Test Plan
1. Make the RNTester app build a separate APK per abi
2. assemble debug and release APK
```shell
$ cd react-native
react-native$ rm -rf ./node_modules && yarn install
react-native$ ./gradlew :ReactAndroid:cleanReactNdkLib :RNTester:android:app:clean :RNTester:android:app:assembleRelease
```
I found these to be beneficial against some build errors:
```shell
$ cd react-native
react-native$ rm -rf ./.buckd/
react-native$ rm -rf ./buck-out/
```
3. use `adb install` to install the `arm64-v8a` APK on a PIXEL 2XL
```shell 
$ cd react-native
react-native$ adb install ./RNTester/android/app/build/outputs/apk/app-arm64-v8a-release.apk
```
4. the app runs fine in both cases.

## Related PRs
[X] ~~Todo: change [instructions](https://github.com/facebook/react-native-website) on building from source.~~ facebook/react-native-website#394

## Release Notes
[ANDROID] [ENHANCEMENT] [ReactAndroid] - add x86_64 arm64-v8a support.

### CODE REVIEW

**Comment by hramos** [comment]
Sweet. Let's make sure `test_android` is passing before merging. At this time, the `test_android` failure is unrelated to your change as `master` is failing with the same reason (#19309).

**Comment by hey99xx** [comment]
@hramos in your other comment in https://github.com/facebook/react-native/issues/2814#issuecomment-392894183 you said you prefer to stick with existing JSC that FB uses. So are you ok with this PR integrating with the new JSC from SoftwareMansion? Aren't those conflicting statements, or did something happen to convince FB that new JSC is finally usable/stable?

**Comment by Minishlink** [comment]
@marafat Very nice!
`test_android` is now passing on master, can you rebase your branch please?

**Comment by marafat** [comment]
@hramos looking at the logs of `test_android` it seems that we need to increase the storage specified on the container running the emulator. Not sure if you agree, but if you do, can you point me where I should do it and maybe suggest a new value.

Also, not sure how to fix `test_ios_e2e` but I will look into it.

Thanks!

**Comment by marafat** [comment]
@hramos `test_android` is passing now. I was misled by the error message, the last commit did solve the issue. 

Will continue to look into `test_ios_e2e`.

**Comment by marafat** [comment]
After some digging, I found that the reason `test_ios_e2e` is failing is currently reported here react-native#19538.

**Comment by marafat** [comment]
@hramos All checks pass here - what is the next step?

**Comment by gengjiawen** [comment]
Excellant pr, but why not import android jsc as a gradle lib instead of npm.

File: jsc-android-manager/package.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "name": "jsc-android-manager",
+  "version": "0.0.1",
+  "private": true,
+  "dependencies": {
+    "jsc-android": "^216113.0.0"
```
Comment (gengjiawen) [inline jsc-android-manager/package.json]: Maybe lock this version by remove `^` to keep it predictable.

Comment (gengjiawen) [reply jsc-android-manager/package.json]: Any special reason not put this deps in root `package.json`

**Comment by marafat** [comment]
@gengjiawen ~ Thanks! I did not want to increase the build time by placing this dep in the root `package.json`. It also felt a bit out of context, as all deps in the root file are related to application code and this would be just for build time.

File: jsc-android-manager/package.json (hunk 1 above)
Comment (marafat) [reply jsc-android-manager/package.json]: Version was locked and updated to latest from [npm/jsc-android]. Test plan succeeded. 🤞 on ci tests

**Comment by marafat** [comment]
One thing I would like a feedback on is about [downloadJSCHeaders](https://github.com/facebook/react-native/blob/master/ReactAndroid/build.gradle#L128) task.

I am not sure how to update this task to pull headers that are in sync with the revision defined in [`jsc-android-manager/package.json`](https://github.com/facebook/react-native/pull/19536/files#diff-9c1eaa2e08b05ef814a52fce4c94e2e5R6), that is 224109.

@hramos - I hope you can help or @mention someone who can.

**Comment by hey99xx** [comment]
@marafat I really want to get this PR done so I'll offer some help.

JSC headers' base url has been changed from https://svn.webkit.org/repository/webkit/!svn/bc/174650/trunk/Source/JavaScriptCore/API/ to 
https://raw.githubusercontent.com/WebKit/webkit/38b15a3ba3c1b0798f2036f7cea36ffdc096202e/Source/JavaScriptCore/API/ in this commit https://github.com/facebook/react-native/commit/5d4b4765834f10ea6463e895084b46f889eb1a48.

svn url was more obvious as it contained the version number 174650 in it. Based on that you can use https://svn.webkit.org/repository/webkit/!svn/bc/224109/trunk/Source/JavaScriptCore/API/

If you look at the commit from the WebKit GitHub mirror https://github.com/WebKit/webkit/commit/38b15a3ba3c1b0798f2036f7cea36ffdc096202e where the new URL is based on, you can see this in the commit message: `git-svn-id: http://svn.webkit.org/repository/webkit/trunk@174650`.

Based on that, we need to find the commit that mirrored 224109, which after few minutes I found to be https://github.com/WebKit/webkit/commit/db7b669da6f28b43c8474d0f881a4ad6e705a75f as it includes `git-svn-id: http://svn.webkit.org/repository/webkit/trunk@224019`.

Then, you should be able to use this as your new base URL: https://raw.githubusercontent.com/WebKit/webkit/db7b669da6f28b43c8474d0f881a4ad6e705a75f/Source/JavaScriptCore/API/

**Comment by hey99xx** [comment]
@marafat I really want to get this PR done so I'll offer some help.

JSC headers' base url has been changed from https://svn.webkit.org/repository/webkit/!svn/bc/174650/trunk/Source/JavaScriptCore/API/ to 
https://raw.githubusercontent.com/WebKit/webkit/38b15a3ba3c1b0798f2036f7cea36ffdc096202e/Source/JavaScriptCore/API/ in this commit https://github.com/facebook/react-native/commit/5d4b4765834f10ea6463e895084b46f889eb1a48.

svn url was more obvious as it contained the version number 174650 in it. So the obvious new base url would be https://svn.webkit.org/repository/webkit/!svn/bc/224109/trunk/Source/JavaScriptCore/API/

**Comment by hey99xx** [comment]
If we still want to use raw.githubusercontent... of WebKit mirror instead of svn.webkit..., maybe @bestander can help as the author of the commit above.

If you look at the commit from the WebKit GitHub mirror https://github.com/WebKit/webkit/commit/38b15a3ba3c1b0798f2036f7cea36ffdc096202e where the new URL is based on, you can see this in the commit message: `git-svn-id: http://svn.webkit.org/repository/webkit/trunk@174650`.

Based on that, we need to find the commit that mirrored 224109, which after few minutes I found to be https://github.com/WebKit/webkit/commit/7d54617067ba6f2333f1b0b46dc39eea9e32a529 as it includes `git-svn-id: http://svn.webkit.org/repository/webkit/trunk@224110` (couldn’t find one that directly matched `224109`).

Then, you should be able to use this as your new base url https://raw.githubusercontent.com/WebKit/webkit/7d54617067ba6f2333f1b0b46dc39eea9e32a529/Source/JavaScriptCore/API/

**Comment by bestander** [comment]
Hey everyone, I am not actively involved with RN recently, working on a product that uses RN now.
I'll point to this PR in our internal group.
Just FYI those things are usually hard to merge, dependencies management is complicated

**Comment by marafat** [comment]
Thank you @bestander for chiming in and for the internal push. I understand your point, and I tried as much as I could to make this PR as slim as possible. 

For my personal notes, I made a document that represents the timeline of issues I faced trying to get this done, and the solution I chose to for each. If you think this would be helpful in speeding up the review process I will be happy to share.

**Comment by marafat** [comment]
Thanks @hey99xx for the detailed help. I will incorporate asap...

**Comment by gengjiawen** [comment]
I think stay in one pr is okay, since the main motivation is getting arm64 support.

File: scripts/android-setup.sh [hunk 2]
```diff
@@ -33,7 +33,7 @@ function getAndroidNDK {
   if [ ! -e $DEPS ]; then
     cd $NDK_HOME
     echo "Downloading NDK..."
-    curl -o ndk.zip https://dl.google.com/android/repository/android-ndk-r10e-linux-x86_64.zip
+    curl -o ndk.zip https://dl.google.com/android/repository/android-ndk-r16-linux-x86_64.zip
```
Comment (gengjiawen) [inline scripts/android-setup.sh]: Remove `android.useDeprecatedNdk=true` in `gradle.properties`

**Comment by bachvanthe1994** [comment]
Hi Mr.Marafat,
I can't see link download this version :( how i can download this version? I need it.

**Comment by patrickkempff** [comment]
@bestander thanks for the update. I am very curious, can you share a bit more information about this? 👍

**Comment by hramos** [comment]
Lots of conflicts here, but they're likely due to the progress that has been made to update several of these dependencies in other PRs.

Like Konstantin said, we're working on making it easier to use a different JSC soon.

**Comment by sospedra** [comment]
Hello dear mantainers <3
This is marked for the milestone 0.58
In order to organize ourselves, is this gonna happen finally? 😱 😍

**Comment by hey99xx** [comment]
@sospedra I think they're just taking this one https://github.com/facebook/react-native/pull/18754. So 64bit binary support will be there, but no update to the JSC engine (like this PR was doing) and will keep using the current android-jsc.

**Comment by hramos** [comment]
Yes, it will happen, though I do not think it will definitely be in place for 0.58. There's some work we need to do to determine what actually needs to be done to support this, and it is being discussed at https://github.com/facebook/react-native/issues/2814. As @hey99xx and @bestander said, we need to do some work related to the JSC used in React Native on Android.

**Comment by hey99xx** [comment]
@hramos Practically everything that this PR was doing is already done by other PRs.

- From https://github.com/facebook/react-native/commit/f3e5cce4745c0ad9a5c697be772757a03e15edc5 "Use new JavaScriptCore from npm" commit message:

> \- Use clang instead of the deprecated gcc
> \- Use libc++ instead of the deprecated gnustl

- 64-bit work was done earlier without the JSC upgrade in https://github.com/facebook/react-native/commit/0a2825f8b3dc183bb17c9aa18d778ddb0f641a66 with a custom 64-bit build of the old android-jsc.aar checked in into the repo.

- NDK bump from 10e to 17b was done in https://github.com/facebook/react-native/commit/6117a6c7205c969f93d39ba02e0583881572d5fa.

- glog and double-conversion updates were done in https://github.com/facebook/react-native/commit/b5fca806059e628edb504cb1bacf62e89ee6f102 and https://github.com/facebook/react-native/commit/f32032e4607b1b269a4d9b6c7efa07f5545ae73e.

Maybe give @marafat chance to look again if there's still anything needed, but I'd think this PR can be closed without being merged in favor of all the other work completed.

**Comment by marafat** [comment]
Closing - Thanks @hey99xx for providing links to all the pieces. I did go over all of them, and I confirm that they all collectively achieve everything this PR is trying to.

@hramos - for those getting here from google, I wonder if you can share a stable release number that will include 64bit support?

**Comment by hramos** [comment]
The new JSC is scheduled for React Native 0.59.

**Pull request closed without merging**